### PR TITLE
Restart long running jobs after compute node visibility change

### PR DESCRIPTION
**Pull request by rossjones**
Compute node visibility will change over time.  This may be caused by network partitions, intentional lack of connectivity, or short term compute node failures (whether h/w or s/w failure). 

When visibility is restored, there is a risk that tasks that should be running will be lost.  There are several options to avoid this happening.

1. Abandon all existing state, launch a fresh instance of the compute and inform the requester node that the previous node-id is no more. The requester node will send work.
2. Similar to 1, but relaunch the same instance, and just inform that requester node that all tasks that were running have failed.
3. Have the compute node determine what to do with tasks, cancelling (as we can't yet record an async failure) any short lived tasks, and cancel/restarting any long-lived tasks.  

This PR implements option 3, but this may not be optimal because:

* The requester node is informed of state changes via the compute node Cancel/Run calls, and these may not happen for a variety of reasons
* The state changes may happen, but be out of sync with the requester node resulting in needing to take more action locally
* We don't know the state of the underlying execution environment, which may result in orphan processes eventually consuming node resources

This PR adds a `Startup` structure to contain tasks that we want to run on the compute node startup (e.g. see other tickets about cleaning up various job inputs).  Here we have added a check on all of the executions that the execution store believes are running, and then based on the job type, we decide on what action to take. 

For short-lived jobs we ideally want to mark them as failed, but unfortunately that is only available as part of the current Run() flow - so instead we use `Cancel()` with the intention to change it later. 

For long-lived jobs we want to be able to check and see if it is running, after which there is nothing to do, or whether we need to re-Run the execution.  Unfortunately we don't have a mechanism yet to tell what the state of the underlying process is (e.g. docker, wasm etc) and so we instead `Cancel()` and `Run()`.  

The solution for neither job is ideal, but it may be possible to extend the pluggable executors to support the desired behaviour.

### CODE REVIEW

File: pkg/compute/startup.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+func (s *Startup) runExecution(ctx context.Context, execution store.LocalExecutionState) error {
+	// We want to ensure this execution is running.  If we just call .Run, there's a
+	// chance we'll end up with two copies running if the previous version is alive
+	// for whatever reason.  We'll call Cancel here so that we can be sure that it
+	// is not running before we ask for it to be run.
+	// TODO: Find a better way of either:
+	// * Findout out whether the underlying process (e.g. docker) is still running, or
+	// * Have .Run be idempotent for the execution id without relying on the store.
```
[the remark this answers is not in the document]
Comment (rossjones) [reply pkg/compute/startup.go:97]: For executors like docker, that can tag an underlying service and then look it up again later, it should be straight-forward.  For those like wasm which are running in the executor plugin process, it's much harder.  We have https://github.com/bacalhau-project/bacalhau/issues/2739 which discusses some options for handling this.

File: pkg/compute/startup.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+	return nil
+}
+
+func (s *Startup) failExecution(ctx context.Context, execution store.LocalExecutionState) error {
+	// Calling cancel with the execute buffer will update our local state, and
+	// then when it calls the underlying baseexecutor, that will inform the requester
+	// node. We really want to _Fail_ the execution, but that's currently only possible
+	// as part of a call to .Run().
+	err := s.execBuffer.Cancel(ctx, execution)
```
[the remark this answers is not in the document]
Comment (rossjones) [reply pkg/compute/startup.go:85]: It's currently captured in #2739 and I think we need to narrow down the states/logic for that issue before we add specific tasks.

File: pkg/compute/store/boltdb/store.go [hunk 3]
```diff
@@ -370,7 +427,30 @@ func (s *Store) updateExecutionState(tx *bolt.Tx, request store.UpdateExecutionS
...
+	// that the index for live executions does not include this ID.
+	var indexError error
+	indexKey := []byte(localExecutionState.Execution.ID)
+	bkt := s.getLiveIndexBucket(tx)
+
+	if localExecutionState.State.IsExecuting() {
+		indexError = bkt.Put(indexKey, []byte{})
+	} else {
+		indexError = bkt.Delete(indexKey)
+	}
```
[the remark this answers is not in the document]
Comment (rossjones) [reply pkg/compute/store/boltdb/store.go:441]: Delete does nothing if the provided key is not found it happily returns nil - and only errors if it is called on a readonly trxn. Put will happily overwrite what is already present for that key and only errors with wrong trxn type, or if the key or value are too large, which is 32k for the key and `1<<31 -1` for the value.

File: pkg/compute/startup.go [hunk 4]
```diff
@@ -65,13 +68,13 @@ func (s *Startup) ensureLiveJobs(ctx context.Context) error {
...
 				if err != nil {
-					return err
+					errs = multierror.Append(errs, err)
 				}
 			}
 		}
 	}
 
-	return nil
+	return errs.Unwrap()
```
Comment (wdbaruni) [inline pkg/compute/startup.go:77]: `Unwrap()` only returns the first error. You want to use `ErrorOrNil()`

https://go.dev/play/p/ev8lpHV3qo3

Comment (rossjones) [reply pkg/compute/startup.go]: That'll teach me to properly read the docs and not assume it works like unwraps elsewhere :(

**Pull request merged**